collection/zset/zset.go

```diff
@@ -0,0 +1,879 @@
+func (z *Float64Set) Contains(value string) bool {
+	_, ok := z.Score(value)
+	return ok
+}
+
+// Score returns the score of the value in the sorted set.
+//
+// Score is the replacement of ZSCORE command of redis.
+func (z *Float64Set) Score(value string) (float64, bool) {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	score, ok := z.dict[value]
+	return score, ok
+}
+
+// Rank returns the rank of element in the sorted set, with the scores
+// ordered from low to high.
+// The rank (or index) is 0-based, which means that the member with the lowest
+// score has rank 0.
+// -1 is returned when value is not found.
+//
+// Rank is the replacement of ZRANK command of redis.
+func (z *Float64Set) Rank(value string) int {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	score, ok := z.dict[value]
+	if !ok {
+		return -1
+	}
+	// NOTE: list.Rank returns 1-based rank.
+	return z.list.Rank(score, value) - 1
+}
+
+// RevRank returns the rank of element in the sorted set, with the scores
+// ordered from high to low.
+// The rank (or index) is 0-based, which means that the member with the highest
+// score has rank 0.
+// -1 is returned when value is not found.
+//
+// RevRank is the replacement of ZREVRANK command of redis.
+func (z *Float64Set) RevRank(value string) int {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	score, ok := z.dict[value]
+	if !ok {
+		return -1
+	}
+	// NOTE: list.Rank returns 1-based rank.
+	return z.list.Rank(score, value) - 1
+}
+
+// Count returns the number of elements in the sorted set at element with a score
+// between min and max (including elements with score equal to min or max).
+//
+// Count is the replacement of ZCOUNT command of redis.
+func (z *Float64Set) Count(min, max float64) int {
+	return z.CountWithOpt(min, max, Float64RangeOpt{})
+}
+
+func (z *Float64Set) CountWithOpt(min, max float64, opt Float64RangeOpt) int {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	first := z.list.FirstInRange(min, max, opt)
+	if first == nil {
+		return 0
+	}
+	// sub 1 for 1-based rank
+	firstRank := z.list.Rank(first.score, first.value) - 1
+	last := z.list.LastInRange(min, max, opt)
+	if last == nil {
+		return z.list.length - firstRank
+	}
+	// sub 1 for 1-based rank
+	lastRank := z.list.Rank(last.score, last.value) - 1
+	return lastRank - firstRank + 1
+}
+
+// Range returns the specified inclusive range of elements in the sorted set by rank(index).
+// Both start and stop are 0-based, they can also be negative numbers indicating
+// offsets from the end of the sorted set, with -1 being the last element of the sorted set,
+// and so on.
+//
+// The returned elements are ordered by score, from lowest to highest.
+// Elements with the same score are ordered lexicographically.
+//
+// This function won't panic even when the given rank out of range.
+//
+// Range is the replacement of ZRANGE command of redis.
+func (z *Float64Set) Range(start, stop int) []Float64Node {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	// Convert negative rank
```

Comment:
One-line comment better be a full sentence.

Reply:
Done.